src/exchange/checkpoint.py

```diff
@@ -1,11 +1,67 @@
+from copy import deepcopy
+from typing import List
 from attrs import define, field
 
 
 @define
 class Checkpoint:
     """Checkpoint that counts the tokens in messages between the start and end index"""
 
-    start_index: int = field(default=-1)
-    end_index: int = field(default=0)
+    start_index: int = field(default=0)  # inclusive
+    end_index: int = field(default=0)  # inclusive
     token_count: int = field(default=0)
-    latest_generated_tokens: int = field(default=0)
+
+    def __copy__(self) -> "Checkpoint":
+        """
+        Returns a deep copy of the Checkpoint object.
+        """
+        return Checkpoint(start_index=self.start_index, end_index=self.end_index, token_count=self.token_count)
+
+    def __deepcopy__(self, _) -> "Checkpoint":
+        """
+        Returns a deep copy of the Checkpoint object.
+        """
+        return Checkpoint(start_index=self.start_index, end_index=self.end_index, token_count=self.token_count)
+
+
+@define
+class CheckpointData:
+    """Aggregates all information about checkpoints"""
+
+    total_token_count: int = field(default=0)
+    checkpoints: List[Checkpoint] = field(default=[])
+    message_index_offset: int = field(default=0)
+
+    def __copy__(self) -> "CheckpointData":
+        """Returns a deep copy of the CheckpointData object."""
+        return CheckpointData(
+            total_token_count=self.total_token_count,
+            checkpoints=deepcopy(self.checkpoints),
```

Comment:
for expectations on behavior, maybe changing this to be a copy() instead of deepcopy() or even a copy(self)

Reply:
actually we don't use the __copy__ methods at all - removing them to avoid confusion.